File: Asgn4/asg4.pl
% /*
%  * CSCI3180 Principles of Programming Languages
%  *
%  * --- Declaration ---
%  *
%  * I declare that the assignment here submitted is original except for source
%  * material explicitly acknowledged. I also acknowledge that I am aware of
%  * University policy and regulations on honesty in academic work, and of the
%  * disciplinary guidelines and procedures applicable to breaches of such policy
%  * and regulations, as contained in the website
%  * http://www.cuhk.edu.hk/policy/academichonesty/
%  *
%  * Assignment 4
%  * Name : Lee Tsz Yan
%  * Student ID : 1155110177
%  * Email Addr : tylee8@cse.cuhk.edu.hk
%  */


% append/3
append([],L,L).
append([X|L1],L2,[X|L3]):-append(L1,L2,L3).

% Q1
% a
element_last(X, L) :- append(_, [X], L).

%element_n(X, L, N) :- append([A|X], [B], L).
% b
element_n(X, [X|_], s(0)).
element_n(X, [_|L], s(N)) :- element_n(X, L, N).

% c
remove_n(X, [X|K], s(0), K).
remove_n(X, [A|L], s(N), [A|K]) :- remove_n(X, L, N, K).

% d
% remove_n(a, L, s(s(0)), [c, d, b, e]).

% e
insert_n(H, L, N, A) :- remove_n(H, A, N, L).

% f
repeat_three([], []).
repeat_three([A|B], [A,A,A|C]) :- repeat_three(B,C).

% g
% repeat_three(X, [i,i,i,m,m,m,n,n,n]).

% sum(0,0,0) :- sum(s(0), s(0), s(s(0))).

% Q2
% sum/3
sum(0,X,X).
sum(s(X),Y,s(Z)) :- sum(X,Y,Z).

% a
% mt(a, [mt(b, [mt(e, []), mt(f, [])]), mt(c, []), mt(d, [mt(g, [])])]).

% b
% mt(_, []).
is_tree(mt(_, T)) :- tree(T).

tree([]).
tree([T|U]) :- is_tree(T), tree(U).

% c
num_node(mt(_,[]), s(0)).
num_node(mt(_, [R|T]), s(N)) :- num_node(R, N1), sub_node(T, N2), sum(N1, N2, N).
sub_node([], 0).
sub_node([R|T], N) :- num_node(R, N1), sub_node(T, N2), sum(N1, N2, N).

% d
% sum_length(mt(_, []), 0).

sum_length(T, L) :- node_count(T, L, 0).
node_count(mt(_, []), 0, _).
node_count(mt(_, [H|T]), S, L) :- node_count(H, N1, s(L)), tree_node(T, N2, s(L)), sum(N1, N2, N), sum(N, s(L), S).
tree_node([], 0, _).
tree_node([H|T], S, L) :- tree_node(T, N1, L), node_count(H, N2, L), sum(N1, N2, N), sum(L, N, S).

File: Asgn4/asg4.ml
(* % /*
%  * CSCI3180 Principles of Programming Languages
%  *
%  * --- Declaration ---
%  *
%  * I declare that the assignment here submitted is original except for source
%  * material explicitly acknowledged. I also acknowledge that I am aware of
%  * University policy and regulations on honesty in academic work, and of the
%  * disciplinary guidelines and procedures applicable to breaches of such policy
%  * and regulations, as contained in the website
%  * http://www.cuhk.edu.hk/policy/academichonesty/
%  *
%  * Assignment 4
%  * Name : Lee Tsz Yan
%  * Student ID : 1155110177
%  * Email Addr : tylee8@cse.cuhk.edu.hk
%  */ *)


datatype suit = Clubs | Diamonds | Hearts | Spades;
datatype hand = Nothing | Pair | Two_Pairs | Three_Of_A_Kind | Full_House | Four_Of_A_Kind | Flush | Straight;
type card = suit * int;
type rank_pair = hand * (int * int)list;

(* fun gt_suit(_:suit, Spades:suit) = false
    | gt_suit(Spades:suit, _:suit) = true
    | gt_suit(_:suit, Clubs:suit) = true
    | gt_suit(Clubs:suit, _:suit) = false
    | gt_suit(s1:suit, s2:suit) = s1 = Hearts andalso s2 = Diamonds; *)

fun last(L:card list):card = 
    if(length(L) = 1)then(hd(L))
    else(last(L));

(* 1. *)
fun check_flush(L:card list):bool = 
    if(length(L) = 1)then(true)
    else((#1(hd(L)) = #1(hd(tl(L)))) andalso check_flush(tl(L)));

(* 2. *)
fun compare_flush(L1:card list, L2:card list):string = 
    if(L1 = nil)then("This is a tie")
    else if(#2(hd(L1)) > #2(hd(L2)))then("Hand 1 wins")
    else if(#2(hd(L1)) < #2(hd(L2)))then("Hand 2 wins")
    (* else if(#1(hd(L1)) <> #1(hd(L2)) andalso gt_suit(#1(hd(L1)), #1(hd(L2))))then("Hand 1 wins")
    else if(#1(hd(L1)) <> #1(hd(L2)))then("Hand 2 wins") *)
    else(compare_flush(tl(L1), tl(L2)));

(* 3. *)
fun check_straight(L:card list):bool = 
    if(length(L) = 1 orelse (length(L) = 2 andalso #2(hd(L)) = 10 andalso #2(hd(tl(L))) = 1))then(true)
    else(#2(hd(L)) - 1 = #2(hd(tl(L))) andalso check_straight(tl(L)));

(* 4. *)
fun compare_straight(L1:card list, L2:card list):string = 
    if(#2(hd(L1)) > #2(hd(L2)))then("Hand 1 wins")
    else if(#2(hd(L1)) < #2(hd(L2)))then("Hand 2 wins")
    else if(#2(hd(L1)) = 13 andalso #2(hd(L2)) = 13 andalso #2(last(L1)) = 1 andalso #2(last(L2)) = 9)then("Hand 1 wins")
    else if(#2(hd(L1)) = 13 andalso #2(hd(L2)) = 13 andalso #2(last(L1)) = 9 andalso #2(last(L2)) = 1)then("Hand 2 wins")
    else("This is a tie");

fun count(L:card list):(int * int)list = 
    let
        fun sub_count(nil, L, i) = (L, i)::nil
        | sub_count((h:card)::t, aCard, i) = 
            if aCard = #2(h) then sub_count(t, #2(h), i+1)
            else(aCard, i)::sub_count(t, #2(h), 1)
    in
        sub_count(tl(L), #2(hd(L)), 1)
    end;

fun sort [] = [] | sort (x::xs) = 
    let
        fun insert(x:(int * int), []) = [x]
            | insert(x:(int * int), y::ys) = 
                if(#2(x) >= #2(y)) then x::y::ys
                else y::insert(x, ys)
    in
        insert(x, sort xs)
    end;

(* 5. *)
fun count_patterns(L:card list):rank_pair = 
    let
        val L2 = sort(count(L))
         
        fun sub_patten(L: (int * int) list as (_, 4)::t, _):rank_pair = (Four_Of_A_Kind, L2)
            | sub_patten((_, 1)::t, 0) = sub_patten(t, 1)
            | sub_patten((_, 2)::t, 0) = sub_patten(t, 2)
            | sub_patten((_, 3)::t, 0) = sub_patten(t, 3)
            | sub_patten((_, 1)::t, 2) = (Pair, L2)
            | sub_patten((_, 2)::t, 2) = (Two_Pairs, L2)
            | sub_patten((_, 1)::t, 3) = (Three_Of_A_Kind, L2)
            | sub_patten((_, 2)::t, 3) = (Full_House, L2)
            | sub_patten(_, _) = (Nothing, L2)
    in
        sub_patten(L2, 0)
    end;

(* 6. *)
fun compare_count(L1:card list, L2:card list):string = 
    (* tie for hand => False, 2nd wins => False, 1st wins => True *)
    let
        fun gt_hand(Nothing, _) = false 
            | gt_hand(_, Four_Of_A_Kind) = false
            | gt_hand(Pair, Pair) = false
            | gt_hand(Pair, Two_Pairs) = false
            | gt_hand(Pair, Three_Of_A_Kind) = false
            | gt_hand(Pair, Full_House) = false
            | gt_hand(Two_Pairs, Two_Pairs) = false
            | gt_hand(Two_Pairs, Three_Of_A_Kind) = false
            | gt_hand(Two_Pairs, Full_House) = false
            | gt_hand(Three_Of_A_Kind, Three_Of_A_Kind) = false
            | gt_hand(Three_Of_A_Kind, Full_House) = false
            | gt_hand(Full_House, Full_House) = false
            | gt_hand(_,_) = true

        val P1 = count_patterns(L1)
        val P2 = count_patterns(L2)
        fun compare(nil, nil) = "This is a tie"
            | compare((h1:(int * int)list), (h2:(int * int)list)) = 
                if (#1(hd(h1)) > #1(hd(h2))) then "Hand 1 wins"
                else if(#1(hd(h1)) < #1(hd(h2))) then "Hand 2 wins"
                else compare(tl(h1), tl(h2))
    in
        if(gt_hand(#1(P1), #1(P2))) then "Hand 1 wins"
        else if(gt_hand(#1(P2), #1(P1))) then "Hand 2 wins"
        else compare(#2(P1), #2(P2))
    end;


